draco/debug.py

```diff
@@ -0,0 +1,218 @@
+        # Nested preference counts
+        pref_count_dict: dict[str, dict[str, int]] = {
+            chart_name: self.__full_pref_count_dict(
+                self.draco, spec, self.feature_names
+            )
+            for chart_name, spec in self.specs.items()
+        }
+        # Flattened preference counts
+        pref_tuples = self.__unnest_pref_count_dict(pref_count_dict)
+        pref_tuples_with_weights = self.__pref_tuples_extended_with_weights(
+            pref_tuples, self.weights
+        )
+        return pd.DataFrame(
+            data=pref_tuples_with_weights,
+            columns=["chart_name", "pref_name", "count", "weight"],
+        )
+
+    @staticmethod
+    def __construct_weight_dict(
+        feature_names: set[str],
+        custom_weights: dict[str, int],
+        default_weights: dict[str, int],
+    ) -> dict[str, int]:
+        default_weight = 0
+        result: dict[str, int] = {}
+        for feature_name in feature_names:
+            key = f"{feature_name}_weight"
+            value = custom_weights.get(key, default_weights.get(key, default_weight))
+            result[key] = value
+        return result
+
+    @staticmethod
+    def __full_pref_count_dict(
+        draco: Draco, spec: str | Iterable[str], feature_names: set[str]
+    ) -> dict[str, int]:
+        """
+        Constructs a dict of preference counts,
+        including all the soft constraint names in the keys
+        """
+        base: dict[str, int] = draco.count_preferences(spec) or {}
+        included_feature_names = set(base.keys())
+        not_included_feature_names = feature_names - included_feature_names
+        return base | {feature_name: 0 for feature_name in not_included_feature_names}
+
+    @staticmethod
+    def __unnest_pref_count_dict(
+        dct: dict[str, dict[str, int]]
+    ) -> list[tuple[str, str, int]]:
+        """Flattens the dict of dicts into a list of tuples"""
+        return [
+            (chart_name, pref_name, count)
+            for chart_name, prefs in dct.items()
+            for pref_name, count in prefs.items()
+        ]
+
+    @staticmethod
+    def __pref_tuple_extended_with_weight(
+        tpl: tuple[str, str, int], weights: dict[str, int]
+    ) -> tuple[str, str, int, int]:
+        """
+        Extends a ``(chart_name, pref_name, count)``
+        tuple to include ``weight`` as its last element.
+        """
+        chart_name, pref_name, count = tpl
+        weight = weights[f"{pref_name}_weight"]
+        return chart_name, pref_name, count, weight
+
+    @staticmethod
+    def __pref_tuples_extended_with_weights(
+        tuples: list[tuple[str, str, int]], weights: dict[str, int]
+    ) -> list[tuple[str, str, int, int]]:
+        cls = DracoDebug
+        return [cls.__pref_tuple_extended_with_weight(tpl, weights) for tpl in tuples]
+
+
+class ChartConfig(NamedTuple):
+    title: str
+    sort_x: alt.Sort | str
+    sort_y: alt.Sort | str
+
+
+class DracoDebugPlotter:
```

Comment:
Any idea how to test this? If not, we should consider leaving the code in the notebook.

Reply:
I added tests in dabb05b checking for structural properties of the produced chart based on the input dataset and configs. What do you think about them?